README.txt
# Easy Code Scanner Unity 3D plugin for Android and iOS
Ready to use plugin to capture barcode and associated data in an Unity3D project.

# Usage

## Installation


### In a new Test Project

* File > New Project > select a folder
* Assets > import package > Custom Package 
* Select the EasyCodeScanner_Unity3D_android_ios_v1.x.unitypackage file then import
* re-open Unity3D with the EasyCodeScanner.unity file located in the folder you created
* File > Build settings
* Select iOS or Android

On iOS
* Build the plugin in Unity3D (switch to iOS and 'Build')
* In the generated XCode project in 'Build Settings'
    - set 'Enable C++ Exception' to YES
    - set Architecture to 'Standard armv7'
* Build and Run with XCode 

On Android
* Build and Run and select a destination APK file name.


### In an existing Project

* Import the package in Unity or copy the content of the package in the 'Assets' folder of your Unity project (Asset > Import Package > Custom Package...).
* Select your platform in Unity3d (File > Build Setting) and switch to the desired platform (android or iOS)

On iOS
* Build the plugin in Unity3D (switch to iOS and 'Build')
* In the generated XCode project in 'Build Settings'
    - set 'Enable C++ Exception' to YES
    - set Architecture to 'Standard armv7'
* Build and Run with XCode 

On Android
* Put the 2 plugin activities (RootActivity & CameraActivity) in your own Manifest.xml file, otherwise make sure the bundle identifier in the player settings is the one declared in Assets/Plugins/Android/AndroidManifest.xml in the player settings (com.c4mprod.ezcodescanner).
  In your Manifest, RootActivity does not need to be MAIN if an other android plugin is already MAIN.
* Build and Run and select a destination APK file name.




## Compilation (Optional)

The plugins are already compiled and ready to be used but you may want to modify it to change the behavior of the plugin.
Here are the instructions:

### Android

* First import the /Assets/Plugins/Android/project folder in Eclipse (right-click and Import)
* Make your modifications if needed.
* Export as a JAR file to Assets/Plugins/Android/EZCodeScanner.jar (Right-click > Export, then select JAR and ONLY export 'src' and 'gen').
* Save and replace
* Build and Run on the Unity3D side.

### iOS
* You can modify the EZCodeScannerViewController.h and EZCodeScannerViewController.mm file in Assets/Plugins/iOS


### Modify the auto-generated GameObject name

if you need to change the name of the object generated then you have to change the following lines and recompile the plugin:
* line 254 in CamaraActivity.java for Android
* line 56 in EZCodeScannerViewController.mm for iPhone
* line 31 in EasyCodeScanner.cs



## Q&A

Q : I got the following error, how to solve it?
Undefined symbols for architecture armv7:
"_launchScannerImpl", referenced from:
RegisterMonoModules() in RegisterMonoModules.o
ld: symbol(s) not found for architecture armv7
clang: error: linker command failed with exit code 1 (use -v to see invocation)

A : Check the following points :
1.The package file are most certainly not at the right location, make sur the "Plugin" folder is right under the "Asset" folder:
+Assets
 -CHANGES.txt
 -EasyCodeScanner.cs
 -EasyCodeScanner.unity
 -EasyCodeScannerExample.cs
 +Plugins
  +Android
  +iOS
 -README.txt
2. In the generated iPhone project you have the EZCodeScannerViewController.mm file under the libraries folder
3. The EZCodeScannerViewController.mm file has the Unity-iPhone Target Membership checked (on the right panel) - very important !


Q : On Android it crashes just after pushing the scan button, what is the problem?
A : Make sure you have declared the 2 activities of the plugin in your Manifest and also the set the permissions (CAMERA & VIBRATE).


Q : Do I Need to manually copy files in the Xcode project?
A : No everything should be automatically copied to the Xcode project by unity3D if the plugin files are at the right location.


Q: Can I decode a provided image with the plugin?
A: Yes on iOS devices only, you can provide a Texture2D or a byte array and a call back will give you back the decoded message.

Q: How can I change the Scan UI?
A: You can change the scan message in the launchScanner method. For graphical modification you need to use 2 different tricks
   - convert your images in int arrays and load them statically in your classes
   - Use Resource.Load() and then access them natively via the external data folder (http://stackoverflow.com/questions/9129086/load-local-asset-bundle-through-www)

Q: As I need a bar code reader, and not a QR Code is there anything that can be done ? 
A: As for Barcode, EasyCodeScanner can read a wide variety of codes. You just have to specify in the type of code 

ZBAR_NONE = 0, < no symbol decoded 
ZBAR_PARTIAL = 1, < intermediate status 
ZBAR_EAN8 = 8, < EAN-8 /
ZBAR_UPCE = 9, < UPC-E 
ZBAR_ISBN10 = 10, < ISBN-10 (from EAN-13). @since 0.4 
ZBAR_UPCA = 12, < UPC-A 
ZBAR_EAN13 = 13, < EAN-13 
ZBAR_ISBN13 = 14, < ISBN-13 (from EAN-13). @since 0.4 
ZBAR_I25 = 25, < Interleaved 2 of 5. @since 0.4 
ZBAR_CODE39 = 39, < Code 39. @since 0.4 
ZBAR_PDF417 = 57, < PDF417. @since 0.6 
ZBAR_QRCODE = 64, < QR Code. @since 0.10 
ZBAR_CODE128 = 128, < Code 128 
ZBAR_SYMBOL = 0x00ff, < mask for base symbol type 
ZBAR_ADDON2 = 0x0200, < 2-digit add-on flag 
ZBAR_ADDON5 = 0x0500, < 5-digit add-on flag 
ZBAR_ADDON = 0x0700, < add-on flag mask 
} zbar_symbol_type_t;

for instance, only Barcode
launchScanner( true, "Scanning...", 13, true);

CHANGES.txt
v1.8
- UI and camera bug fixes
- Double scan fix when heavy process in OnScannerMessage

v1.7
- UI bug fix with orientation is set to landscape in Unity3D

v1.6
- Double scan crash fix
- portrait and landscape support on iOS

v1.5
- EasyCodeScanner API is now in a separate C# script for better clarity
- Easier integration : bridge GameObject is automatically created, no more prefab to include
- [iOS] New methods to get the scanned picture : 
	public static byte[] getScannerImage();
	public static Texture2D getScannerImage(int texWidth, int texHeigh);
- [iOS] New methods to decode a picture :
	public static void decodeImage(int symbols, byte[] image);
	public static void decodeImage(int symbols, Texture2D texture)
	OnDecoderMessage Action callback when image is decoded

v1.4
- New callback "onScannerEvent" to indicate the opening ("EVENT_OPENED") and closing ("EVENT_CLOSED") of the plugin 

v1.3
- "GameObject" becomes "CodeScannerBridge", you need to rename the GameObject to make it work.
- Prefab fix

v1.2
- [Android]Improved code detection
- [Android]Optimized detection area
- [Android]Faster camera preview
- [Android]Better front camera preview quality
- Action "OnScannerSucced" becomes "OnScannerSuccess"


v1.1
- [Android] Icon link bug fix in the Manifest
- [Android] Now compatible with other concurrent Android plugins (static call to scanner)
- Added event Action to be notified in other classes

v1.0
- First release


EasyCodeScanner.cs
using UnityEngine;
using System.Collections;
using System.Runtime.InteropServices;
using System;

public class EasyCodeScanner : MonoBehaviour{
	
	//Action
	public static event Action<string> OnScannerMessage;
	public static event Action<string> OnScannerEvent;
	public static event Action<string> OnDecoderMessage;
	
	#if UNITY_IPHONE
		public struct ConfigStruct
		{
			public bool showUI;
	    	public string defaultText;
			public int symbols;
	    	public bool forceLandscape;
		}
		[DllImport ("__Internal")] private static extern void launchScannerImpl(ref ConfigStruct conf);
		[DllImport ("__Internal")] private static extern bool getScannedImageImpl(ref IntPtr pixelBytes, ref int imageDataLength);
		[DllImport ("__Internal")] private static extern void decodeImageImpl(int symbols, byte[] pixelBytes, long length);
	#endif
	
	static EasyCodeScanner instance;
	public static void Initialize()
	{
		if (instance == null)
		{
			GameObject newGameObject = new GameObject("CodeScannerBridge");
			newGameObject.AddComponent<EasyCodeScanner>();
			instance = newGameObject.GetComponent<EasyCodeScanner>();
		}
	}
	
	void Start()
	{
		
	}
	
	//Callback from Java or Objective-C which returns the data (barcode, url, ...)
	void onScannerMessage(string data){
		//Debug.Log("EasyCodeScanner - onScannerMessage data=:"+data);
		if (OnScannerMessage != null)
        {
            OnScannerMessage(data);
        }
	}
	
	//Callback from Java or Objective-C which notifies an event
	//param : "EVENT_OPENED", "EVENT_CLOSED"
	void onScannerEvent(string eventStr){
		//Debug.Log("EasyCodeScanner - onScannerEvent eventStr=:"+eventStr);
		if (OnScannerEvent != null)
        {
            OnScannerEvent(eventStr);
        }
	}
	
	void onDecoderMessage(string code){
		//Debug.Log("EasyCodeScanner - onDecoderMessage code=:"+code);
		if (OnDecoderMessage != null)
        {
            OnDecoderMessage(code);
        }
	}
	
	public static void launchScanner(bool showUI, string defaultTxt, int symbol, bool forceLandscape) {
		
		if (instance==null) {
			Debug.LogError("EasyCodeScanner - launchScanner error : scanner must be initialized before.");
			return;
		}
		#if UNITY_IPHONE	
			//IPHONE - Display the UIViewController
			ConfigStruct conf = new ConfigStruct();
			conf.showUI = showUI;
			conf.defaultText = defaultTxt;
			conf.symbols = symbol;
			conf.forceLandscape = forceLandscape;
			launchScannerImpl(ref conf);
		#endif
		
		#if UNITY_ANDROID
			//ANDROID - Launch the Activity with an Intent
			AndroidJavaClass ajc = new AndroidJavaClass("com.unity3d.player.UnityPlayer");
			AndroidJavaObject ajo = ajc.GetStatic<AndroidJavaObject>("currentActivity");
		
			//In case RootActivity is the MAIN in the manifest
			//ajo.Call("launchScannerImpl", showUI, defaultTxt, symbol, forceLandscape);
		
			//In case RootActivity is not the MAIN activity in the manifest (multi-plugin)
			var jc = new AndroidJavaClass("com.c4mprod.ezcodescanner.RootActivity");
    		jc.CallStatic("launchScannerImpl", ajo, showUI, defaultTxt, symbol, forceLandscape);
		#endif
	
	}
	
	public static byte[] getScannerImage() {
		
		if (instance==null) {
			Debug.LogError("EasyCodeScanner - launchScanner error : scanner must be initialized before.");
			return null;
		}
		#if UNITY_IPHONE
			//Get the image from the scanner
			byte[] pixelBytes = null;
			IntPtr ptrImageData = IntPtr.Zero;
	        int imageDataLength = 0;
	        bool success = getScannedImageImpl(ref ptrImageData, ref imageDataLength);
	        if (success)
	        	{
	            	// Load the results into a managed array.
	            	pixelBytes = new byte[imageDataLength];
	            	Marshal.Copy(ptrImageData
	                	, pixelBytes
	                	, 0
	                	, imageDataLength);
	        	}
			return pixelBytes;
		#elif UNITY_ANDROID
			//Not yet implemented
			return null;
		#else
			return null;
		#endif
	}
	
	public static Texture2D getScannerImage(int texWidth, int texHeigh) {
		
		if (instance==null) {
			Debug.LogError("EasyCodeScanner - launchScanner error : scanner must be initialized before.");
			return null;
		}
		#if UNITY_IPHONE
			byte[] pixelBytes = getScannerImage();
			//Convert the image to a texture
			Texture2D tex = new Texture2D(texWidth, texHeigh);
        	tex.LoadImage(pixelBytes);
			return tex;
		#elif UNITY_ANDROID
			//Not yet implemented
			return null;
		#else
			return null;
		#endif
	}
	
	//Calls OnDecoderMessage in return
	public static void decodeImage(int symbols, byte[] image) {
		
		if (instance==null) {
			Debug.LogError("EasyCodeScanner - launchScanner error : scanner must be initialized before.");
			return;
		}
		#if UNITY_IPHONE
			decodeImageImpl(symbols, image, image.Length);
		#elif UNITY_ANDROID
			//Not yet implemented
			return;
		#else
			return;
		#endif
	}
	
	//Calls OnDecoderMessage in return
	public static void decodeImage(int symbols, Texture2D texture) {
		
		if (instance==null) {
			Debug.LogError("EasyCodeScanner - launchScanner error : scanner must be initialized before.");
			return;
		}
		#if UNITY_IPHONE
			byte[] pixelBytes = texture.EncodeToPNG();
			decodeImageImpl(symbols, pixelBytes, pixelBytes.Length);
		#elif UNITY_ANDROID
			//Not yet implemented
			return;
		#else
			return;
		#endif
	}
	
}

Plugins/Android/project/src/com/c4mprod/ezcodescanner/RootActivity.java
package com.c4mprod.ezcodescanner;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

import com.c4mprod.ezcodescanner.camera.CameraActivity;
import com.unity3d.player.UnityPlayerActivity;

public class RootActivity extends UnityPlayerActivity {

    public static final String EXTRA_BOOLEAN_UI        = "show_ui";
    public static final String EXTRA_STRING_TXT        = "default_text";
    public static final String EXTRA_INT_SYMBOLS       = "symbols_mask";
    public static final String EXTRA_BOOLEAN_LANDSCAPE = "force_landscape";

    /**
     * Called when the activity is first created.
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    /**
     * Launch the scanner view with specific config if root activity is the MAIN activity (no other plugins)
     * 
     * @param showUI
     *            display the ui overlay
     * @param defaultText
     *            The default text to show while scanning
     * @param symbols
     *            Mask with values in Symbol class, for instance Symbol.QRCODE or (Symbol.QRCODE|Symbol.EAN13). -1 for all symbols.
     */
    public void launchScannerImpl(boolean showUI, String defaultText, int symbols, boolean forceLandscape) {

        Intent intent = new Intent(this, CameraActivity.class);
        intent.putExtra(EXTRA_BOOLEAN_UI, showUI);
        intent.putExtra(EXTRA_STRING_TXT, defaultText);
        intent.putExtra(EXTRA_INT_SYMBOLS, symbols);
        intent.putExtra(EXTRA_BOOLEAN_LANDSCAPE, forceLandscape);
        startActivity(intent);
    }

    /**
     * Launch the scanner statically from an existing activity view with specific config (Other activity than RootActivity is MAIN - other plugin).
     * 
     * @param showUI
     *            display the ui overlay
     * @param defaultText
     *            The default text to show while scanning
     * @param symbols
     *            Mask with values in Symbol class, for instance Symbol.QRCODE or (Symbol.QRCODE|Symbol.EAN13). -1 for all symbols.
     */
    public static void launchScannerImpl(Activity root, boolean showUI, String defaultText, int symbols, boolean forceLandscape) {

        Intent intent = new Intent(root, CameraActivity.class);
        intent.putExtra(RootActivity.EXTRA_BOOLEAN_UI, showUI);
        intent.putExtra(RootActivity.EXTRA_STRING_TXT, defaultText);
        intent.putExtra(RootActivity.EXTRA_INT_SYMBOLS, symbols);
        intent.putExtra(RootActivity.EXTRA_BOOLEAN_LANDSCAPE, forceLandscape);
        root.startActivity(intent);
    }

}

Plugins/Android/project/src/com/c4mprod/ezcodescanner/views/CameraUI.java
package com.c4mprod.ezcodescanner.views;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.View;

public final class CameraUI extends View {

    private final Paint        paint;
    private static final int[] SCANNER_ALPHA  = { 0, 64, 128, 192, 255, 192, 128, 64 };
    private final int          laserColor;
    private int                scannerAlpha;
    private final int          frameColor;
    private final int          maskColor;
    private Rect               mFrame;
    int                        mWidth;
    int                        mHeight;

    private static final int   PERCENT_LEFT   = 20;
    private static final int   PERCENT_TOP    = 30;
    private static final int   PERCENT_RIGHT  = 20;
    private static final int   PERCENT_BOTTOM = 30;

    public CameraUI(Context context) {
        super(context);

        // Initialize these once for performance rather than calling them every time in onDraw().
        paint = new Paint();
        laserColor = 0xFFFF0000;
        frameColor = 0xCC000000;
        maskColor = 0x90000000;

    }

    @SuppressLint("DrawAllocation")
    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {

        if (changed) {
            mWidth = right - left;
            mHeight = bottom - top;

            if (mWidth < mHeight) {
                // portrait
                mFrame = new Rect((mWidth * PERCENT_LEFT) / 100, (mHeight * PERCENT_TOP) / 100, mWidth - (mWidth * PERCENT_RIGHT) / 100,
                                  mHeight - (mHeight * PERCENT_BOTTOM) / 100);
            } else {
                // landscape
                mFrame = new Rect((mWidth * PERCENT_TOP) / 100, (mHeight * PERCENT_RIGHT) / 100, mWidth - (mWidth * PERCENT_BOTTOM) / 100,
                                  mHeight - (mHeight * PERCENT_LEFT) / 100);
            }
        }

        super.onLayout(changed, left, top, right, bottom);
    }

    public Rect getCaptureFrame() {
        return mFrame;
    }

    @Override
    public void onDraw(Canvas canvas) {

        if (mFrame != null) {
            int width = canvas.getWidth();
            int height = canvas.getHeight();

            // Draw the exterior (i.e. outside the framing rect) darkened
            paint.setColor(maskColor);
            canvas.drawRect(0, 0, width, mFrame.top, paint);
            canvas.drawRect(0, mFrame.top, mFrame.left, mFrame.bottom + 1, paint);
            canvas.drawRect(mFrame.right + 1, mFrame.top, width, mFrame.bottom + 1, paint);
            canvas.drawRect(0, mFrame.bottom + 1, width, height, paint);

            // Draw a two pixel solid black border inside the framing rect
            paint.setColor(frameColor);
            canvas.drawRect(mFrame.left, mFrame.top, mFrame.right + 1, mFrame.top + 2, paint);
            canvas.drawRect(mFrame.left, mFrame.top + 2, mFrame.left + 2, mFrame.bottom - 1, paint);
            canvas.drawRect(mFrame.right - 1, mFrame.top, mFrame.right + 1, mFrame.bottom - 1, paint);
            canvas.drawRect(mFrame.left, mFrame.bottom - 1, mFrame.right + 1, mFrame.bottom + 1, paint);

            // Draw a red "laser scanner" line through the middle to show decoding
            // is active
            paint.setColor(laserColor);
            paint.setAlpha(SCANNER_ALPHA[scannerAlpha]);
            scannerAlpha = (scannerAlpha + 1) % SCANNER_ALPHA.length;
            int middle = mFrame.height() / 2 + mFrame.top;
            canvas.drawRect(mFrame.left + 2, middle - 1, mFrame.right - 1, middle + 2, paint);

            postInvalidate();
        }
    }

}


Plugins/Android/project/src/com/c4mprod/ezcodescanner/camera/CameraPreview.java
package com.c4mprod.ezcodescanner.camera;

import java.util.List;

import android.content.Context;
import android.hardware.Camera;
import android.hardware.Camera.AutoFocusCallback;
import android.hardware.Camera.PreviewCallback;
import android.hardware.Camera.Size;
import android.util.Log;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

/** A basic Camera preview class */
public class CameraPreview extends SurfaceView implements SurfaceHolder.Callback {
    private final SurfaceHolder     mHolder;
    private final Camera            mCamera;
    private final PreviewCallback   previewCallback;
    private final AutoFocusCallback autoFocusCallback;
    private final boolean           isLandscape;

    public CameraPreview(Context context, Camera camera, PreviewCallback previewCb, AutoFocusCallback autoFocusCb, boolean land) {
        super(context);
        mCamera = camera;
        previewCallback = previewCb;
        autoFocusCallback = autoFocusCb;
        isLandscape = land;

        /*
         * Set camera to continuous focus if supported, otherwise use
         * software auto-focus. Only works for API level >=9.
         */

        // Camera.Parameters parameters = camera.getParameters();
        // for (String f : parameters.getSupportedFocusModes()) {
        // if (f == Parameters.FOCUS_MODE_CONTINUOUS_PICTURE) {
        // mCamera.setFocusMode(Parameters.FOCUS_MODE_CONTINUOUS_PICTURE);
        // autoFocusCallback = null;
        // break;
        // }
        // }

        // Install a SurfaceHolder.Callback so we get notified when the
        // underlying surface is created and destroyed.
        mHolder = getHolder();
        mHolder.addCallback(this);

        // deprecated setting, but required on Android versions prior to 3.0
        mHolder.setType(SurfaceHolder.SURFACE_TYPE_PUSH_BUFFERS);
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        // The Surface has been created, now tell the camera where to draw the preview.
        try {
            mCamera.setPreviewDisplay(holder);
        } catch (Exception e) {
            Log.d("DBG", "Error setting camera preview: " + e.getMessage());
        }
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        // Camera preview released in activity
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        /*
         * If your preview can change or rotate, take care of those events here.
         * Make sure to stop the preview before resizing or reformatting it.
         */
        if (mHolder.getSurface() == null) {
            // preview surface does not exist
            return;
        }

        // stop preview before making changes
        try {
            mCamera.stopPreview();
        } catch (Exception e) {
            // ignore: tried to stop a non-existent preview
        }

        try {
            // Hard code camera surface rotation 90 degs to match Activity view in portrait
            if (!isLandscape) {
                mCamera.setDisplayOrientation(90);
            }
            mCamera.setPreviewDisplay(mHolder);
            mCamera.setPreviewCallback(previewCallback);

            if (isLandscape) {
                Size size = getLargestSupportedPreviewSize(mCamera);
                if (size != null) {
                    Camera.Parameters parameters = mCamera.getParameters();
                    parameters.setPreviewSize(size.width, size.height);
                    mCamera.setParameters(parameters);
                }
            }
            mCamera.startPreview();
            mCamera.autoFocus(autoFocusCallback);
        } catch (Exception e) {
            Log.d("DBG", "Error starting camera preview: " + e.getMessage());
        }
    }

    public Size getLargestSupportedPreviewSize(Camera camera) {
        Camera.Parameters parameters = mCamera.getParameters();
        List<Size> sizes = parameters.getSupportedPreviewSizes();
        int maxvalue = 0;
        Size size = null;
        for (Size s : sizes) {
            // Log.d("DEBUG", "available size=" + s.width + "x" + s.height);
            if (s.width > maxvalue) {
                maxvalue = s.width;
                size = s;
            }
        }
        // Log.d("DEBUG", "=>Choosen size=" + size.width + "x" + size.height);
        return size;
    }
}


EasyCodeScannerExample.cs
using UnityEngine;
using System.Collections;
using System.Runtime.InteropServices;
using System;

public class EasyCodeScannerExample : MonoBehaviour {
	
	static string dataStr;
	public Renderer PlaneRender;
	
	void Start () {
		dataStr = "";
		
		// Initialize EasyCodeScanner
		EasyCodeScanner.Initialize();
		
		//Register on Actions
		EasyCodeScanner.OnScannerMessage += onScannerMessage;
		EasyCodeScanner.OnScannerEvent += onScannerEvent;
		EasyCodeScanner.OnDecoderMessage += onDecoderMessage;
		
		//Screen.orientation = ScreenOrientation.LandscapeLeft;
	}
	
	void OnDestroy() {
		
		//Unregister
		EasyCodeScanner.OnScannerMessage -= onScannerMessage;
		EasyCodeScanner.OnScannerEvent -= onScannerEvent;
		EasyCodeScanner.OnDecoderMessage -= onDecoderMessage;
	}
	
	public void Update() {
			
		if (Input.GetKeyDown(KeyCode.Escape)) { 
			Application.Quit();
		}
		
	}
	
	void OnGUI ()
	{
		//--- To be replaced by your controls, just to showcase ---
		if (GUI.Button(new Rect (10, 125, 300, 100), "Launch scanner"))
		{
			//for code list, see : http://zbar.sourceforge.net/api/zbar_8h.html#f7818ad6458f9f40362eecda97acdcb0
			// and zbar.h 
			/*
			typedef enum zbar_symbol_type_e {
		    ZBAR_NONE        =      0,  < no symbol decoded 
		    ZBAR_PARTIAL     =      1,  < intermediate status 
		    ZBAR_EAN8        =      8,  < EAN-8 /
		    ZBAR_UPCE        =      9,  < UPC-E 
		    ZBAR_ISBN10      =     10,  < ISBN-10 (from EAN-13). @since 0.4 
		    ZBAR_UPCA        =     12,  < UPC-A 
		    ZBAR_EAN13       =     13,  < EAN-13 
		    ZBAR_ISBN13      =     14,  < ISBN-13 (from EAN-13). @since 0.4 
		    ZBAR_I25         =     25,  < Interleaved 2 of 5. @since 0.4 
		    ZBAR_CODE39      =     39,  < Code 39. @since 0.4 
		    ZBAR_PDF417      =     57,  < PDF417. @since 0.6 
		    ZBAR_QRCODE      =     64,  < QR Code. @since 0.10 
		    ZBAR_CODE128     =    128,  < Code 128 
		    ZBAR_SYMBOL      = 0x00ff,  < mask for base symbol type 
		    ZBAR_ADDON2      = 0x0200,  < 2-digit add-on flag 
		    ZBAR_ADDON5      = 0x0500,  < 5-digit add-on flag 
		    ZBAR_ADDON       = 0x0700,  < add-on flag mask 
			} zbar_symbol_type_t;
			*/
			//for instance, only QRCodes
			//launchScanner( true, "Scanning...", 64, true);
			
			//all codes at the same time : value = -1
			EasyCodeScanner.launchScanner( true, "Scanning...", -1, true);
		}
	
		dataStr = GUI.TextArea(new Rect (10, 275, 300, 100), dataStr);
		//---
	}
	
	//Callback when returns from the scanner
	void onScannerMessage(string data){
		Debug.Log("EasyCodeScannerExample - onScannerMessage data=:"+data);
		dataStr = data;
		
		//Just to show case : get the image and display it on a Plane
		Texture2D tex = EasyCodeScanner.getScannerImage(200, 200);
		PlaneRender.material.mainTexture = tex;
		
		//Just to show case : decode a texture/image - refer to code list
		EasyCodeScanner.decodeImage(-1, tex);
	}
	
	//Callback which notifies an event
	//param : "EVENT_OPENED", "EVENT_CLOSED"
	void onScannerEvent(string eventStr){
		Debug.Log("EasyCodeScannerExample - onScannerEvent:"+eventStr);
	}
	
	//Callback when decodeImage has decoded the image/texture 
	void onDecoderMessage(string data){
		Debug.Log("EasyCodeScannerExample - onDecoderMessage data:"+data);
		dataStr = data;
	}
	
	
	
}